### PR TITLE
[no-release-notes] Added document explaining how branch control works

### PR DESCRIPTION
**Pull request by Hydrocharged**
This adds rate-limiting to the branch control tables, such that we'll only write to disc once every 30 seconds. This only applies to the server context. In addition, we save when closing the server, in case there are any outstanding changes.

I've also added a doc that goes over how the new tree works, so give it a look over so I know that it's actually explaining everything in a helpful way.

### CODE REVIEW

**Comment by timsehn** [comment]
What is we took something like the journal approach. Write everything in an append only log and memory. Flush memory to disk every 30s and make sure the disk matches the log. Process the log on server start up.

**Comment by Hydrocharged** [comment]
You're commenting too quickly, I can't reply in time lol. I was initially going to have a `defer` in our `main` function so that we may still save in the event of panics, but decided not to as I didn't think it would be too well received. Plus a hard crash would skip a panic and probably corrupt the data anyway.

Rather than redesigning the persistence file (the current implementation can still read old files just fine), I could add a timer that just checks if we've written anything new every 30 seconds. That way you have to crash within a 30 second window.

**Comment by Hydrocharged** [comment]
Refer to my [previous PR](https://github.com/dolthub/dolt/pull/5238#issuecomment-1408750719) for additional performance details. I marked my `.dolt` folder safe in Windows Defender, and got a massive speed increase, but the slowdown in seconds is the same so the old results are still valid. The second column is the current PR, and the third column is with all branch control features disabled (matching, updating, and writing). The times are basically the same, so branch control is no longer a performance bottleneck in any way.

| Branches | PR | BC Disabled |
| --- | --- | --- |
| 1-1000 | 2.9997827s | 2.8207606s |
| 1001-2000 | 3.4695718s | 3.0173117s |
| 2001-3000 | 3.3305172s | 3.0130936s |
| 3001-4000 | 3.4147117s | 3.0546962s |
| 4001-5000 | 3.412387s | 3.0660925s |
| 5001-6000 | 3.4887713s | 3.1632063s |
| 6001-7000 | 3.5722639s | 3.2130576s |
| 7001-8000 | 3.5816457s | 3.2778602s |
| 8001-9000 | 3.7440973s | 3.3437402s |
| 9001-10000 | 3.7729238s | 3.3980096s |
| 10001-11000 | 3.7556886s | 3.3885819s |
| 11001-12000 | 3.8091551s | 3.6122785s |
| 12001-13000 | 3.8943095s | 3.516991s |
| 13001-14000 | 3.8938174s | 3.5630588s |
| 14001-15000 | 3.9194065s | 3.5986724s |
| 15001-16000 | 4.0206627s | 3.6218522s |
| 16001-17000 | 4.1716665s | 3.673915s |
| 17001-18000 | 4.1324553s | 3.7357573s |
| 18001-19000 | 4.674165s | 3.9311572s |
| 19001-20000 | 4.3136222s | 3.8883617s |
| 20001-21000 | 4.4060064s | 4.0044791s |
| 21001-22000 | 4.4535748s | 3.9289011s |
| 22001-23000 | 4.4088215s | 4.0417361s |
| 23001-24000 | 4.5111226s | 4.1004445s |
| 24001-25000 | 4.4945715s | 4.1782415s |
| 25001-26000 | 4.7150272s | 4.1905258s |
| 26001-27000 | 4.8273291s | 4.2364116s |
| 27001-28000 | 4.6173375s | 4.2617925s |
| 28001-29000 | 4.7799583s | 4.3209953s |
| 29001-30000 | 4.7376641s | 4.3783441s |
| 30001-31000 | 4.8465451s | 4.4292506s |
| 31001-32000 | 4.9283657s | 4.9240565s |
| 32001-33000 | 4.9706314s | 4.5746606s |
| 33001-34000 | 5.0174291s | 4.6025678s |
| 34001-35000 | 5.0641859s | 4.6450646s |
| 35001-36000 | 5.2275941s | 4.7232923s |
| 36001-37000 | 5.2029828s | 4.7498093s |
| 37001-38000 | 5.1917467s | 4.7750036s |
| 38001-39000 | 5.4292168s | 4.9341868s |
| 39001-40000 | 5.3113481s | 4.940087s |
| 40001-41000 | 6.4393686s | 4.9614712s |
| 41001-42000 | 5.4832574s | 5.0646487s |
| 42001-43000 | 5.4973566s | 5.080836s |
| 43001-44000 | 5.5340903s | 5.1101539s |
| 44001-45000 | 5.6216834s | 5.397475s |
| 45001-46000 | 5.7306536s | 5.2362396s |
| 46001-47000 | 5.8645589s | 5.2724064s |
| 47001-48000 | 5.8164563s | 5.2956509s |
| 48001-49000 | 6.0186221s | 5.448082s |
| 49001-50000 | 6.0395162s | 5.495015s |
| 50001-51000 | 5.9332769s | 5.4240265s |
| 51001-52000 | 6.4667944s | 5.6305223s |
| 52001-53000 | 6.1776507s | 5.5627534s |
| 53001-54000 | 6.0520629s | 5.5479444s |
| 54001-55000 | 5.6911835s | 5.2218556s |
| 55001-56000 | 6.1504864s | 5.7007487s |
| 56001-57000 | 6.4182645s | 5.7120638s |
| 57001-58000 | 6.3216872s | 5.782022s |
| 58001-59000 | 6.5874909s | 5.8446721s |
| 59001-60000 | 6.3337953s | 6.0256877s |
| 60001-61000 | 6.2308633s | 5.924013s |
| 61001-62000 | 6.088585s | 5.7531279s |
| 62001-63000 | 5.8957495s | 5.5300166s |
| 63001-64000 | 5.9523008s | 5.6577093s |
| 64001-65000 | 6.1836349s | 6.3999397s |
| 65001-66000 | 6.7689368s | 6.3711401s |
| 66001-67000 | 6.7518726s | 6.4436791s |
| 67001-68000 | 7.1686975s | 6.2970035s |
| 68001-69000 | 5.8968536s | 5.7245103s |
| 69001-70000 | 6.0359227s | 5.8441587s |
| 70001-71000 | 6.3019371s | 5.8838892s |
| 71001-72000 | 5.864275s | 5.9225545s |
| 72001-73000 | 6.1006435s | 7.3162753s |
| 73001-74000 | 6.6567813s | 6.5104176s |
| 74001-75000 | 7.4333914s | 7.0049516s |
| 75001-76000 | 7.3773512s | 7.0104539s |
| 76001-77000 | 7.4420375s | 7.0688195s |
| 77001-78000 | 7.5882041s | 7.1478569s |
| 78001-79000 | 7.5663849s | 7.3418859s |
| 79001-80000 | 7.7945532s | 7.4968857s |
| 80001-81000 | 7.6016397s | 7.3363382s |
| 81001-82000 | 7.0825516s | 7.0478001s |
| 82001-83000 | 6.5220959s | 6.4239608s |
| 83001-84000 | 6.6209009s | 6.4485163s |
| 84001-85000 | 6.647223s | 6.6607712s |
| 85001-86000 | 7.023605s | 6.5961636s |
| 86001-87000 | 6.7519849s | 6.6200182s |
| 87001-88000 | 6.9726148s | 6.7513892s |
| 88001-89000 | 7.0218485s | 6.7963283s |
| 89001-90000 | 7.2247612s | 7.07148s |
| 90001-91000 | 7.0124156s | 7.1745012s |
| 91001-92000 | 7.0661394s | 6.8954279s |
| 92001-93000 | 7.2517761s | 7.183778s |
| 93001-94000 | 7.5186902s | 7.3831025s |
| 94001-95000 | 8.6974101s | 8.5290713s |
| 95001-96000 | 8.7923444s | 8.4820027s |
| 96001-97000 | 7.3627085s | 7.5027304s |
| 97001-98000 | 7.3391818s | 7.1729941s |
| 98001-99000 | 7.4928835s | 7.2264133s |
| 99001-100000 | 7.7978012s | 7.4859117s |

File: go/libraries/doltcore/branch_control/expr_parser_node.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
+This document is to show the implementation theory behind the `MatchNode` structure in `expr_parser_node.go`.
```
Comment (zachmu) [inline go/libraries/doltcore/branch_control/expr_parser_node.md:1]: This is a good doc to have, thanks for writing.

No real comments, pretty well explained and clears up some points of confusion I had.

**Review by zachmu — COMMENTED** [review]
The doc is great, let's keep that.

The rate-limited writing isn't worth it. Better to have slower write behavior in the case of lots of inserts than to play games with persistence like this. If writes are too slow, let's iterate to make the storage mechanism more performant, not limit the write interval.

**Comment by Hydrocharged** [comment]
Removed the rate-limiting functionality and I'm just committing the doc.

**Pull request merged**